src/dbnode/persist/fs/registry.go

```diff
@@ -0,0 +1,335 @@
+		r.entries[pathAbs] = status
+	}
+
+	return status, nil
+}
+
+type fileStatus struct {
+	sync.RWMutex
+	status    FileStatus
+	pathAbs   string
+	watchable watch.Watchable
+	refs      []*fileRef
+
+	nowFn clock.NowFn
+
+	logger *zap.Logger
+}
+
+func newFileStatus(pathAbs string, nowFn clock.NowFn, logger *zap.Logger) *fileStatus {
+	f := &fileStatus{
+		status: FileStatus{
+			Type: AvailableFileStatusType,
+		},
+		pathAbs: pathAbs,
+		nowFn:   nowFn,
+		logger:  logger,
+	}
+	watchable := watch.NewWatchable()
+	watchable.Update(f.status)
+	f.watchable = watchable
+	return f
+}
+
+func (f *fileStatus) newFileRef(ref *os.File) *fileRef {
+	fileRef := newFileRef(f, ref)
+	f.Lock()
+	f.refs = append(f.refs, fileRef)
+	f.Unlock()
+	return fileRef
+}
+
+func (f *fileStatus) removeFileRef(elem *fileRef) error {
+	f.Lock()
+	defer f.Unlock()
+
+	if len(f.refs) == 0 {
+		return fmt.Errorf("remove file ref not found for: %v", elem.PathAbs())
+	}
+
+	refs := make([]*fileRef, 0, len(f.refs)-1)
+	for _, ptr := range f.refs {
+		if ptr != elem {
+			refs = append(refs, ptr)
+		}
+	}
+	if len(refs) != len(f.refs)-1 {
+		return fmt.Errorf("remove file ref not found for: %v", elem.PathAbs())
+	}
+
+	f.refs = refs
+	return nil
+}
+
+func (f *fileStatus) Status() FileStatus {
+	f.RLock()
+	v := f.status
+	f.RUnlock()
+	return v
+}
+
+func (f *fileStatus) NotifyAndRemove() error {
+	f.setStatus(FileStatus{Type: RemovingFileStatusType})
+	defer func() {
+		f.setStatus(FileStatus{Type: RemovedFileStatusType})
+	}()
+
+	timeout := f.nowFn().Add(notifyAndRemoveTimeout)
+	for f.nowFn().Before(timeout) && f.numRefs() > 0 {
+		time.Sleep(notifyAndRemoveCheckInterval)
+	}
+
+	if f.numRefs() > 0 {
+		f.logger.Error("timed out for delayed remove of file",
+			zap.String("path", f.pathAbs))
+	}
+
+	return os.Remove(f.pathAbs)
+}
+
+func (f *fileStatus) setStatus(v FileStatus) {
+	f.Lock()
+	f.status = v
+	f.watchable.Update(v)
+	f.Unlock()
+}
+
+func (f *fileStatus) numRefs() int {
+	f.RLock()
+	v := len(f.refs)
+	f.RUnlock()
+	return v
+}
+
+func (f *fileStatus) Open() (File, error) {
+	ref, err := os.Open(f.pathAbs)
+	if err != nil {
+		return nil, err
+	}
+	return f.newFileRef(ref), nil
+}
+
+func (f *fileStatus) OpenFile(flag int, perm os.FileMode) (File, error) {
+	ref, err := os.OpenFile(f.pathAbs, flag, perm)
+	if err != nil {
+		return nil, err
+	}
+	return f.newFileRef(ref), nil
+}
+
+// Ensure fileRef implements File
+var _ File = &fileRef{}
+
+type fileRef struct {
+	file *fileStatus
+	ref  *os.File
+}
+
+func newFileRef(file *fileStatus, ref *os.File) *fileRef {
+	return &fileRef{file: file, ref: ref}
+}
+
+func (f *fileRef) Status() FileStatus {
+	return f.file.Status()
+}
+
+func (f *fileRef) PathAbs() string {
+	return f.file.pathAbs
+}
+
+func (f *fileRef) WatchStatus(watcher FileStatusWatcher) close.SimpleCloser {
+	_, watch, _ := f.file.watchable.Watch()
+
+	// Spawn a new goroutine that will terminate when the
```

Comment:
should the lifecycle be controllable directly? like what if i want to open a file, watch for some status changes and then stop watching (w/o removing)